File: modules/sandbox/ext/libriscv/examples/webapi/cpp-httplib/modules/sandbox/ext/godot-cpp/src/variant/aabb.cpp
/**************************************************************************/
/*  aabb.cpp                                                              */
/**************************************************************************/
/*                         This file is part of:                          */
/*                             GODOT ENGINE                               */
/*                        https://godotengine.org                         */
/**************************************************************************/
/* Copyright (c) 2014-present Godot Engine contributors (see AUTHORS.md). */
/* Copyright (c) 2007-2014 Juan Linietsky, Ariel Manzur.                  */
/*                                                                        */
/* Permission is hereby granted, free of charge, to any person obtaining  */
/* a copy of this software and associated documentation files (the        */
/* "Software"), to deal in the Software without restriction, including    */
/* without limitation the rights to use, copy, modify, merge, publish,    */
/* distribute, sublicense, and/or sell copies of the Software, and to     */
/* permit persons to whom the Software is furnished to do so, subject to  */
/* the following conditions:                                              */
/*                                                                        */
/* The above copyright notice and this permission notice shall be         */
/* included in all copies or substantial portions of the Software.        */
/*                                                                        */
/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
/**************************************************************************/

#include <godot_cpp/variant/aabb.hpp>

#include <godot_cpp/variant/string.hpp>
#include <godot_cpp/variant/variant.hpp>

namespace godot {

real_t AABB::get_volume() const {
	return size.x * size.y * size.z;
}

bool AABB::operator==(const AABB &p_rval) const {
	return ((position == p_rval.position) && (size == p_rval.size));
}

bool AABB::operator!=(const AABB &p_rval) const {
	return ((position != p_rval.position) || (size != p_rval.size));
}

void AABB::merge_with(const AABB &p_aabb) {
#ifdef MATH_CHECKS
	if (unlikely(size.x < 0 || size.y < 0 || size.z < 0 || p_aabb.size.x < 0 || p_aabb.size.y < 0 || p_aabb.size.z < 0)) {
		ERR_PRINT("AABB size is negative, this is not supported. Use AABB.abs() to get an AABB with a positive size.");
	}
#endif
	Vector3 beg_1, beg_2;
	Vector3 end_1, end_2;
	Vector3 min, max;

	beg_1 = position;
	beg_2 = p_aabb.position;
	end_1 = size + beg_1;
	end_2 = p_aabb.size + beg_2;

	min.x = (beg_1.x < beg_2.x) ? beg_1.x : beg_2.x;
	min.y = (beg_1.y < beg_2.y) ? beg_1.y : beg_2.y;
	min.z = (beg_1.z < beg_2.z) ? beg_1.z : beg_2.z;

	max.x = (end_1.x > end_2.x) ? end_1.x : end_2.x;
	max.y = (end_1.y > end_2.y) ? end_1.y : end_2.y;
	max.z = (end_1.z > end_2.z) ? end_1.z : end_2.z;

	position = min;
	size = max - min;
}

bool AABB::is_equal_approx(const AABB &p_aabb) const {
	return position.is_equal_approx(p_aabb.position) && size.is_equal_approx(p_aabb.size);
}

bool AABB::is_finite() const {
	return position.is_finite() && size.is_finite();
}

AABB AABB::intersection(const AABB &p_aabb) const {
#ifdef MATH_CHECKS
	if (unlikely(size.x < 0 || size.y < 0 || size.z < 0 || p_aabb.size.x < 0 || p_aabb.size.y < 0 || p_aabb.size.z < 0)) {
		ERR_PRINT("AABB size is negative, this is not supported. Use AABB.abs() to get an AABB with a positive size.");
	}
#endif
	Vector3 src_min = position;
	Vector3 src_max = position + size;
	Vector3 dst_min = p_aabb.position;
	Vector3 dst_max = p_aabb.position + p_aabb.size;

	Vector3 min, max;

	if (src_min.x > dst_max.x || src_max.x < dst_min.x) {
		return AABB();
	} else {
		min.x = (src_min.x > dst_min.x) ? src_min.x : dst_min.x;
		max.x = (src_max.x < dst_max.x) ? src_max.x : dst_max.x;
	}

	if (src_min.y > dst_max.y || src_max.y < dst_min.y) {
		return AABB();
	} else {
		min.y = (src_min.y > dst_min.y) ? src_min.y : dst_min.y;
		max.y = (src_max.y < dst_max.y) ? src_max.y : dst_max.y;
	}

	if (src_min.z > dst_max.z || src_max.z < dst_min.z) {
		return AABB();
	} else {
		min.z = (src_min.z > dst_min.z) ? src_min.z : dst_min.z;
		max.z = (src_max.z < dst_max.z) ? src_max.z : dst_max.z;
	}

	return AABB(min, max - min);
}

// Note that this routine returns the BACKTRACKED (i.e. behind the ray origin)
// intersection point + normal if INSIDE the AABB.
// The caller can therefore decide when INSIDE whether to use the
// backtracked intersection, or use p_from as the intersection, and
// carry on progressing without e.g. reflecting against the normal.
bool AABB::find_intersects_ray(const Vector3 &p_from, const Vector3 &p_dir, bool &r_inside, Vector3 *r_intersection_point, Vector3 *r_normal) const {
#ifdef MATH_CHECKS
	if (unlikely(size.x < 0 || size.y < 0 || size.z < 0)) {
		ERR_PRINT("AABB size is negative, this is not supported. Use AABB.abs() to get an AABB with a positive size.");
	}
#endif
	Vector3 end = position + size;
	real_t tmin = -1e20;
	real_t tmax = 1e20;
	int axis = 0;

	// Make sure r_inside is always initialized,
	// to prevent reading uninitialized data in the client code.
	r_inside = false;

	for (int i = 0; i < 3; i++) {
		if (p_dir[i] == 0) {
			if ((p_from[i] < position[i]) || (p_from[i] > end[i])) {
				return false;
			}
		} else { // ray not parallel to planes in this direction
			real_t t1 = (position[i] - p_from[i]) / p_dir[i];
			real_t t2 = (end[i] - p_from[i]) / p_dir[i];

			if (t1 > t2) {
				SWAP(t1, t2);
			}
			if (t1 >= tmin) {
				tmin = t1;
				axis = i;
			}
			if (t2 < tmax) {
				if (t2 < 0) {
					return false;
				}
				tmax = t2;
			}
			if (tmin > tmax) {
				return false;
			}
		}
	}

	// Did the ray start from inside the box?
	// In which case the intersection returned is the point of entry
	// (behind the ray start) or the calling routine can use the ray origin as intersection point.
	r_inside = tmin < 0;

	if (r_intersection_point) {
		*r_intersection_point = p_from + p_dir * tmin;

		// Prevent float error by making sure the point is exactly
		// on the AABB border on the relevant axis.
		r_intersection_point->coord[axis] = (p_dir[axis] >= 0) ? position.coord[axis] : end.coord[axis];
	}
	if (r_normal) {
		*r_normal = Vector3();
		(*r_normal)[axis] = (p_dir[axis] >= 0) ? -1 : 1;
	}

	return true;
}

bool AABB::intersects_segment(const Vector3 &p_from, const Vector3 &p_to, Vector3 *r_intersection_point, Vector3 *r_normal) const {
#ifdef MATH_CHECKS
	if (unlikely(size.x < 0 || size.y < 0 || size.z < 0)) {
		ERR_PRINT("AABB size is negative, this is not supported. Use AABB.abs() to get an AABB with a positive size.");
	}
#endif
	real_t min = 0, max = 1;
	int axis = 0;
	real_t sign = 0;

	for (int i = 0; i < 3; i++) {
		real_t seg_from = p_from[i];
		real_t seg_to = p_to[i];
		real_t box_begin = position[i];
		real_t box_end = box_begin + size[i];
		real_t cmin, cmax;
		real_t csign;

		if (seg_from < seg_to) {
			if (seg_from > box_end || seg_to < box_begin) {
				return false;
			}
			real_t length = seg_to - seg_from;
			cmin = (seg_from < box_begin) ? ((box_begin - seg_from) / length) : 0;
			cmax = (seg_to > box_end) ? ((box_end - seg_from) / length) : 1;
			csign = -1.0;

		} else {
			if (seg_to > box_end || seg_from < box_begin) {
				return false;
			}
			real_t length = seg_to - seg_from;
			cmin = (seg_from > box_end) ? (box_end - seg_from) / length : 0;
			cmax = (seg_to < box_begin) ? (box_begin - seg_from) / length : 1;
			csign = 1.0;
		}

		if (cmin > min) {
			min = cmin;
			axis = i;
			sign = csign;
		}
		if (cmax < max) {
			max = cmax;
		}
		if (max < min) {
			return false;
		}
	}

	Vector3 rel = p_to - p_from;

	if (r_normal) {
		Vector3 normal;
		normal[axis] = sign;
		*r_normal = normal;
	}

	if (r_intersection_point) {
		*r_intersection_point = p_from + rel * min;
	}

	return true;
}

bool AABB::intersects_plane(const Plane &p_plane) const {
	Vector3 points[8] = {
		Vector3(position.x, position.y, position.z),
		Vector3(position.x, position.y, position.z + size.z),
		Vector3(position.x, position.y + size.y, position.z),
		Vector3(position.x, position.y + size.y, position.z + size.z),
		Vector3(position.x + size.x, position.y, position.z),
		Vector3(position.x + size.x, position.y, position.z + size.z),
		Vector3(position.x + size.x, position.y + size.y, position.z),
		Vector3(position.x + size.x, position.y + size.y, position.z + size.z),
	};

	bool over = false;
	bool under = false;

	for (int i = 0; i < 8; i++) {
		if (p_plane.distance_to(points[i]) > 0) {
			over = true;
		} else {
			under = true;
		}
	}

	return under && over;
}

Vector3 AABB::get_longest_axis() const {
	Vector3 axis(1, 0, 0);
	real_t max_size = size.x;

	if (size.y > max_size) {
		axis = Vector3(0, 1, 0);
		max_size = size.y;
	}

	if (size.z > max_size) {
		axis = Vector3(0, 0, 1);
	}

	return axis;
}

int AABB::get_longest_axis_index() const {
	int axis = 0;
	real_t max_size = size.x;

	if (size.y > max_size) {
		axis = 1;
		max_size = size.y;
	}

	if (size.z > max_size) {
		axis = 2;
	}

	return axis;
}

Vector3 AABB::get_shortest_axis() const {
	Vector3 axis(1, 0, 0);
	real_t min_size = size.x;

	if (size.y < min_size) {
		axis = Vector3(0, 1, 0);
		min_size = size.y;
	}

	if (size.z < min_size) {
		axis = Vector3(0, 0, 1);
	}

	return axis;
}

int AABB::get_shortest_axis_index() const {
	int axis = 0;
	real_t min_size = size.x;

	if (size.y < min_size) {
		axis = 1;
		min_size = size.y;
	}

	if (size.z < min_size) {
		axis = 2;
	}

	return axis;
}

AABB AABB::merge(const AABB &p_with) const {
	AABB aabb = *this;
	aabb.merge_with(p_with);
	return aabb;
}

AABB AABB::expand(const Vector3 &p_vector) const {
	AABB aabb = *this;
	aabb.expand_to(p_vector);
	return aabb;
}

AABB AABB::grow(real_t p_by) const {
	AABB aabb = *this;
	aabb.grow_by(p_by);
	return aabb;
}

void AABB::get_edge(int p_edge, Vector3 &r_from, Vector3 &r_to) const {
	ERR_FAIL_INDEX(p_edge, 12);
	switch (p_edge) {
		case 0: {
			r_from = Vector3(position.x + size.x, position.y, position.z);
			r_to = Vector3(position.x, position.y, position.z);
		} break;
		case 1: {
			r_from = Vector3(position.x + size.x, position.y, position.z + size.z);
			r_to = Vector3(position.x + size.x, position.y, position.z);
		} break;
		case 2: {
			r_from = Vector3(position.x, position.y, position.z + size.z);
			r_to = Vector3(position.x + size.x, position.y, position.z + size.z);

		} break;
		case 3: {
			r_from = Vector3(position.x, position.y, position.z);
			r_to = Vector3(position.x, position.y, position.z + size.z);

		} break;
		case 4: {
			r_from = Vector3(position.x, position.y + size.y, position.z);
			r_to = Vector3(position.x + size.x, position.y + size.y, position.z);
		} break;
		case 5: {
			r_from = Vector3(position.x + size.x, position.y + size.y, position.z);
			r_to = Vector3(position.x + size.x, position.y + size.y, position.z + size.z);
		} break;
		case 6: {
			r_from = Vector3(position.x + size.x, position.y + size.y, position.z + size.z);
			r_to = Vector3(position.x, position.y + size.y, position.z + size.z);

		} break;
		case 7: {
			r_from = Vector3(position.x, position.y + size.y, position.z + size.z);
			r_to = Vector3(position.x, position.y + size.y, position.z);

		} break;
		case 8: {
			r_from = Vector3(position.x, position.y, position.z + size.z);
			r_to = Vector3(position.x, position.y + size.y, position.z + size.z);

		} break;
		case 9: {
			r_from = Vector3(position.x, position.y, position.z);
			r_to = Vector3(position.x, position.y + size.y, position.z);

		} break;
		case 10: {
			r_from = Vector3(position.x + size.x, position.y, position.z);
			r_to = Vector3(position.x + size.x, position.y + size.y, position.z);

		} break;
		case 11: {
			r_from = Vector3(position.x + size.x, position.y, position.z + size.z);
			r_to = Vector3(position.x + size.x, position.y + size.y, position.z + size.z);

		} break;
	}
}

Variant AABB::intersects_segment_bind(const Vector3 &p_from, const Vector3 &p_to) const {
	Vector3 inters;
	if (intersects_segment(p_from, p_to, &inters)) {
		return inters;
	}
	return Variant();
}

Variant AABB::intersects_ray_bind(const Vector3 &p_from, const Vector3 &p_dir) const {
	Vector3 inters;
	bool inside = false;

	if (find_intersects_ray(p_from, p_dir, inside, &inters)) {
		// When inside the intersection point may be BEHIND the ray,
		// so for general use we return the ray origin.
		if (inside) {
			return p_from;
		}

		return inters;
	}
	return Variant();
}

AABB::operator String() const {
	return "[P: " + position.operator String() + ", S: " + size + "]";
}

} // namespace godot

File: modules/cross_runtime/NET/utilities/Data_Types/Aabb.cs
using System;
using System.Diagnostics.CodeAnalysis;
using System.Globalization;
using System.Runtime.InteropServices;

#nullable enable
namespace Godot
{
[Serializable]
[StructLayout(LayoutKind.Sequential)]
public struct AABB : IEquatable<AABB>
{
    public enum Axis
    {
        X = 0,
        Y = 1,
        Z = 2
    }

    private Vector3 _position;
    private Vector3 _size;

    public Vector3 Position
    {
        readonly get { return _position; }
        set { _position = value; }
    }

    public Vector3 Size
    {
        readonly get { return _size; }
        set { _size = value; }
    }

    public Vector3 End
    {
        readonly get { return Add(_position, _size); }
        set { _size = Sub(value, _position); }
    }

    public readonly float Volume
    {
        get { return _size.X * _size.Y * _size.Z; }
    }

    public readonly AABB Abs()
    {
        Vector3 end = End;
        Vector3 topLeft = Min(_position, end);
        return new AABB(topLeft, AbsVector(_size));
    }

    public readonly Vector3 GetCenter()
    {
        return Add(_position, Mul(_size, 0.5f));
    }

    public readonly bool Encloses(AABB with)
    {
        Vector3 srcMin = _position;
        Vector3 srcMax = Add(_position, _size);
        Vector3 dstMin = with._position;
        Vector3 dstMax = Add(with._position, with._size);

        return srcMin.X <= dstMin.X &&
               srcMax.X >= dstMax.X &&
               srcMin.Y <= dstMin.Y &&
               srcMax.Y >= dstMax.Y &&
               srcMin.Z <= dstMin.Z &&
               srcMax.Z >= dstMax.Z;
    }

    public readonly AABB Expand(Vector3 point)
    {
        Vector3 begin = _position;
        Vector3 end = Add(_position, _size);

        if (point.X < begin.X) begin.X = point.X;
        if (point.Y < begin.Y) begin.Y = point.Y;
        if (point.Z < begin.Z) begin.Z = point.Z;

        if (point.X > end.X) end.X = point.X;
        if (point.Y > end.Y) end.Y = point.Y;
        if (point.Z > end.Z) end.Z = point.Z;

        return new AABB(begin, Sub(end, begin));
    }

    public readonly Vector3 GetEndpoint(int idx)
    {
        switch (idx)
        {
            case 0: return new Vector3(_position.X, _position.Y, _position.Z);
            case 1: return new Vector3(_position.X, _position.Y, _position.Z + _size.Z);
            case 2: return new Vector3(_position.X, _position.Y + _size.Y, _position.Z);
            case 3: return new Vector3(_position.X, _position.Y + _size.Y, _position.Z + _size.Z);
            case 4: return new Vector3(_position.X + _size.X, _position.Y, _position.Z);
            case 5: return new Vector3(_position.X + _size.X, _position.Y, _position.Z + _size.Z);
            case 6: return new Vector3(_position.X + _size.X, _position.Y + _size.Y, _position.Z);
            case 7: return new Vector3(_position.X + _size.X, _position.Y + _size.Y, _position.Z + _size.Z);
            default:
                throw new ArgumentOutOfRangeException(nameof(idx),
                    $"Index is {idx}, but a value from 0 to 7 is expected.");
        }
    }

    public readonly Vector3 GetLongestAxis()
    {
        Vector3 axis = new Vector3(1f, 0f, 0f);
        float maxSize = _size.X;

        if (_size.Y > maxSize)
        {
            axis = new Vector3(0f, 1f, 0f);
            maxSize = _size.Y;
        }

        if (_size.Z > maxSize)
        {
            axis = new Vector3(0f, 0f, 1f);
        }

        return axis;
    }

    public readonly Axis GetLongestAxisIndex()
    {
        Axis axis = Axis.X;
        float maxSize = _size.X;

        if (_size.Y > maxSize)
        {
            axis = Axis.Y;
            maxSize = _size.Y;
        }

        if (_size.Z > maxSize)
        {
            axis = Axis.Z;
        }

        return axis;
    }

    public readonly float GetLongestAxisSize()
    {
        float maxSize = _size.X;

        if (_size.Y > maxSize)
            maxSize = _size.Y;

        if (_size.Z > maxSize)
            maxSize = _size.Z;

        return maxSize;
    }

    public readonly Vector3 GetShortestAxis()
    {
        Vector3 axis = new Vector3(1f, 0f, 0f);
        float minSize = _size.X;

        if (_size.Y < minSize)
        {
            axis = new Vector3(0f, 1f, 0f);
            minSize = _size.Y;
        }

        if (_size.Z < minSize)
        {
            axis = new Vector3(0f, 0f, 1f);
        }

        return axis;
    }

    public readonly Axis GetShortestAxisIndex()
    {
        Axis axis = Axis.X;
        float minSize = _size.X;

        if (_size.Y < minSize)
        {
            axis = Axis.Y;
            minSize = _size.Y;
        }

        if (_size.Z < minSize)
        {
            axis = Axis.Z;
        }

        return axis;
    }

    public readonly float GetShortestAxisSize()
    {
        float minSize = _size.X;

        if (_size.Y < minSize)
            minSize = _size.Y;

        if (_size.Z < minSize)
            minSize = _size.Z;

        return minSize;
    }

    public readonly Vector3 GetSupport(Vector3 dir)
    {
        Vector3 support = _position;

        if (dir.X > 0.0f) support.X += _size.X;
        if (dir.Y > 0.0f) support.Y += _size.Y;
        if (dir.Z > 0.0f) support.Z += _size.Z;

        return support;
    }

    public readonly AABB Grow(float by)
    {
        AABB res = this;

        res._position.X -= by;
        res._position.Y -= by;
        res._position.Z -= by;
        res._size.X += 2.0f * by;
        res._size.Y += 2.0f * by;
        res._size.Z += 2.0f * by;

        return res;
    }

    public readonly bool HasPoint(Vector3 point)
    {
        if (point.X < _position.X) return false;
        if (point.Y < _position.Y) return false;
        if (point.Z < _position.Z) return false;
        if (point.X > _position.X + _size.X) return false;
        if (point.Y > _position.Y + _size.Y) return false;
        if (point.Z > _position.Z + _size.Z) return false;

        return true;
    }

    public readonly bool HasSurface()
    {
        return _size.X > 0.0f || _size.Y > 0.0f || _size.Z > 0.0f;
    }

    public readonly bool HasVolume()
    {
        return _size.X > 0.0f && _size.Y > 0.0f && _size.Z > 0.0f;
    }

    public readonly AABB Intersection(AABB with)
    {
        Vector3 srcMin = _position;
        Vector3 srcMax = Add(_position, _size);
        Vector3 dstMin = with._position;
        Vector3 dstMax = Add(with._position, with._size);

        Vector3 min;
        Vector3 max;

        if (srcMin.X > dstMax.X || srcMax.X < dstMin.X)
            return new AABB();

        min.X = srcMin.X > dstMin.X ? srcMin.X : dstMin.X;
        max.X = srcMax.X < dstMax.X ? srcMax.X : dstMax.X;

        if (srcMin.Y > dstMax.Y || srcMax.Y < dstMin.Y)
            return new AABB();

        min.Y = srcMin.Y > dstMin.Y ? srcMin.Y : dstMin.Y;
        max.Y = srcMax.Y < dstMax.Y ? srcMax.Y : dstMax.Y;

        if (srcMin.Z > dstMax.Z || srcMax.Z < dstMin.Z)
            return new AABB();

        min.Z = srcMin.Z > dstMin.Z ? srcMin.Z : dstMin.Z;
        max.Z = srcMax.Z < dstMax.Z ? srcMax.Z : dstMax.Z;

        return new AABB(min, Sub(max, min));
    }

    public readonly bool Intersects(AABB with)
    {
        if (_position.X >= with._position.X + with._size.X) return false;
        if (_position.X + _size.X <= with._position.X) return false;
        if (_position.Y >= with._position.Y + with._size.Y) return false;
        if (_position.Y + _size.Y <= with._position.Y) return false;
        if (_position.Z >= with._position.Z + with._size.Z) return false;
        if (_position.Z + _size.Z <= with._position.Z) return false;

        return true;
    }

    public readonly bool IntersectsPlane(Plane plane)
    {
        Vector3 p0 = new Vector3(_position.X, _position.Y, _position.Z);
        Vector3 p1 = new Vector3(_position.X, _position.Y, _position.Z + _size.Z);
        Vector3 p2 = new Vector3(_position.X, _position.Y + _size.Y, _position.Z);
        Vector3 p3 = new Vector3(_position.X, _position.Y + _size.Y, _position.Z + _size.Z);
        Vector3 p4 = new Vector3(_position.X + _size.X, _position.Y, _position.Z);
        Vector3 p5 = new Vector3(_position.X + _size.X, _position.Y, _position.Z + _size.Z);
        Vector3 p6 = new Vector3(_position.X + _size.X, _position.Y + _size.Y, _position.Z);
        Vector3 p7 = new Vector3(_position.X + _size.X, _position.Y + _size.Y, _position.Z + _size.Z);

        bool over = false;
        bool under = false;

        Vector3[] points = [p0, p1, p2, p3, p4, p5, p6, p7];

        for (int i = 0; i < 8; i++)
        {
            float distance = PlaneDistanceTo(plane, points[i]);
            if (distance > 0.0f)
                over = true;
            else
                under = true;
        }

        return under && over;
    }

    public readonly bool IntersectsSegment(Vector3 from, Vector3 to)
    {
        float min = 0.0f;
        float max = 1.0f;

        for (int i = 0; i < 3; i++)
        {
            float segFrom = GetAxis(from, i);
            float segTo = GetAxis(to, i);
            float boxBegin = GetAxis(_position, i);
            float boxEnd = boxBegin + GetAxis(_size, i);
            float cmin;
            float cmax;

            if (segFrom < segTo)
            {
                if (segFrom > boxEnd || segTo < boxBegin)
                    return false;

                float length = segTo - segFrom;
                cmin = segFrom < boxBegin ? (boxBegin - segFrom) / length : 0.0f;
                cmax = segTo > boxEnd ? (boxEnd - segFrom) / length : 1.0f;
            }
            else
            {
                if (segTo > boxEnd || segFrom < boxBegin)
                    return false;

                float length = segTo - segFrom;
                cmin = segFrom > boxEnd ? (boxEnd - segFrom) / length : 0.0f;
                cmax = segTo < boxBegin ? (boxBegin - segFrom) / length : 1.0f;
            }

            if (cmin > min)
                min = cmin;

            if (cmax < max)
                max = cmax;

            if (max < min)
                return false;
        }

        return true;
    }

    public readonly bool IsFinite()
    {
        return IsFinite(_position.X) && IsFinite(_position.Y) && IsFinite(_position.Z) &&
               IsFinite(_size.X) && IsFinite(_size.Y) && IsFinite(_size.Z);
    }

    public readonly AABB Merge(AABB with)
    {
        Vector3 beg1 = _position;
        Vector3 beg2 = with._position;
        Vector3 end1 = Add(beg1, _size);
        Vector3 end2 = Add(beg2, with._size);

        Vector3 min = new Vector3(
            beg1.X < beg2.X ? beg1.X : beg2.X,
            beg1.Y < beg2.Y ? beg1.Y : beg2.Y,
            beg1.Z < beg2.Z ? beg1.Z : beg2.Z
        );

        Vector3 max = new Vector3(
            end1.X > end2.X ? end1.X : end2.X,
            end1.Y > end2.Y ? end1.Y : end2.Y,
            end1.Z > end2.Z ? end1.Z : end2.Z
        );

        return new AABB(min, Sub(max, min));
    }

    public AABB(Vector3 position, Vector3 size)
    {
        _position = position;
        _size = size;
    }

    public AABB(float x, float y, float z, Vector3 size)
    {
        _position = new Vector3(x, y, z);
        _size = size;
    }

    public AABB(float x, float y, float z, float width, float height, float depth)
    {
        _position = new Vector3(x, y, z);
        _size = new Vector3(width, height, depth);
    }

    public static bool operator ==(AABB left, AABB right) => left.Equals(right);
    public static bool operator !=(AABB left, AABB right) => !left.Equals(right);

    public override readonly bool Equals([NotNullWhen(true)] object? obj)
    {
        return obj is AABB other && Equals(other);
    }

    public readonly bool Equals(AABB other)
    {
        return _position.X == other._position.X &&
               _position.Y == other._position.Y &&
               _position.Z == other._position.Z &&
               _size.X == other._size.X &&
               _size.Y == other._size.Y &&
               _size.Z == other._size.Z;
    }

    public override readonly int GetHashCode()
    {
        return HashCode.Combine(
            _position.X, _position.Y, _position.Z,
            _size.X, _size.Y, _size.Z
        );
    }

    public override readonly string ToString() => ToString(null);

    public readonly string ToString(string? format)
    {
        string fmt = string.IsNullOrEmpty(format) ? "G" : format;

        return string.Create(CultureInfo.InvariantCulture, $"{FormatVec3(_position, fmt)}, {FormatVec3(_size, fmt)}");
    }

    private static Vector3 Add(Vector3 a, Vector3 b)
    {
        return new Vector3(a.X + b.X, a.Y + b.Y, a.Z + b.Z);
    }

    private static Vector3 Sub(Vector3 a, Vector3 b)
    {
        return new Vector3(a.X - b.X, a.Y - b.Y, a.Z - b.Z);
    }

    private static Vector3 Mul(Vector3 a, float b)
    {
        return new Vector3(a.X * b, a.Y * b, a.Z * b);
    }

    private static Vector3 Min(Vector3 a, Vector3 b)
    {
        return new Vector3(
            a.X < b.X ? a.X : b.X,
            a.Y < b.Y ? a.Y : b.Y,
            a.Z < b.Z ? a.Z : b.Z
        );
    }

    private static Vector3 AbsVector(Vector3 v)
    {
        return new Vector3(MathF.Abs(v.X), MathF.Abs(v.Y), MathF.Abs(v.Z));
    }

    private static bool IsFinite(float value)
    {
        return !float.IsNaN(value) && !float.IsInfinity(value);
    }

    private static float GetAxis(Vector3 v, int axis)
    {
        return axis switch
        {
            0 => v.X,
            1 => v.Y,
            2 => v.Z,
            _ => throw new ArgumentOutOfRangeException(nameof(axis))
        };
    }

    private static float PlaneDistanceTo(Plane plane, Vector3 point)
    {
        return plane.Normal.X * point.X +
               plane.Normal.Y * point.Y +
               plane.Normal.Z * point.Z +
               plane.D;
    }

    private static string FormatVec3(Vector3 v, string format)
    {
        return "(" +
               v.X.ToString(format, CultureInfo.InvariantCulture) + ", " +
               v.Y.ToString(format, CultureInfo.InvariantCulture) + ", " +
               v.Z.ToString(format, CultureInfo.InvariantCulture) + ")";
    }
}
}